Add Dockerfile

AurelienGasser opened the pull request:
- Follow-up to https://github.com/SubstraFoundation/substra-backend/pull/154
- Companion PR: https://github.com/SubstraFoundation/substra/pull/123

### 1. Build `substra` image

```bash
~/code/substra$ make docker
docker build -f docker/Dockerfile . -t substra:latest
Step 1/9 : FROM python:3.7
[...]
Successfully tagged substra:latest
```

Note: You can specify the tag to build: `make docker DOCKER_TAG=my-version`

### 2. Build `substra-tests` image

```bash
~/code/substra-tests$ make docker
docker build -f docker/Dockerfile . -t substra-tests:latest --build-arg SUBSTRA_VERSION=latest
Step 1/10 : ARG SUBSTRA_VERSION=latest
[...]
Successfully tagged substra-tests:latest
```

Note: 
- You can specify the tag to build: `make docker DOCKER_TAG=my-version`
- You can specify the base `substra` image tag to use: `make docker DOCKER_TAG_SUBSTRA=my-version`

### 3. Run the tests

```bash
$ docker run -it substra-tests:latest bash
root@36dbccf952d0:/usr/src/app/substra-tests# pytest -rs -q tests/test_execution.py -k test_tuples_execution_on_same_node
.                                                                                                                                                                                                    [100%]
1 passed, 10 deselected in 12.55s
```

samlesu commented:
Thanks @AurelienGasser, looks very good.

I'm just wondering, why depending on the substra image? We could install the substra library using pip. It would make the process to build this image simpler.

ClementGautier commented on docker/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+ARG SUBSTRA_VERSION=latest
+FROM substra:$SUBSTRA_VERSION
```
I would start from python and install from pip if it was me so that you don't depend on an intermediary image. Also you could use /usr/src/app as workdir as it would not be taken already.

ClementGautier commented on docker/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+COPY ./*.yaml /usr/src/app/substra-tests/
+COPY ./Makefile /usr/src/app/substra-tests/
+COPY ./tests /usr/src/app/substra-tests/tests/
+COPY ./substratest /usr/src/app/substra-tests/substratest/
```
`COPY . .`, use a .dockerignore if needed

AurelienGasser replied:
Fixed https://github.com/SubstraFoundation/substra-tests/pull/81/commits/328db3dda0e80b1c74c476487f890352bb68bf90

ClementGautier commented on docker/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+COPY ./tests /usr/src/app/substra-tests/tests/
+COPY ./substratest /usr/src/app/substra-tests/substratest/
+
+WORKDIR /usr/src/app/substra-tests
```
Put the workdir on top, it's free cache layer and simplify the reading

AurelienGasser replied:
Fixed https://github.com/SubstraFoundation/substra-tests/pull/81/commits/328db3dda0e80b1c74c476487f890352bb68bf90

ClementGautier commented on docker/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+COPY ./substratest /usr/src/app/substra-tests/substratest/
+
+WORKDIR /usr/src/app/substra-tests
+RUN pip3 install -r requirements.txt
```
to be placed before the source copy to benefit from cache layers.

AurelienGasser replied:
Fixed https://github.com/SubstraFoundation/substra-tests/pull/81/commits/328db3dda0e80b1c74c476487f890352bb68bf90

AurelienGasser commented:
@samlesu @ClementGautier The reason not to install substra from pip is that currently pip only contains publicly released versions. In order for a pip version to be created, and if I understood correctly, two conditions must be present:

1. a git tag must be pushed
2. `substra/__version__.py` needs to be bumped

As of writing this, the latest substra version on pip is over 2 months old. 

Since this PR is about testing prior to releases, depending on a release creates a circular dependency!

Another reason not to use pip is that it allows us to use custom/forked version of substra, which might be useful later. Although this is speculative so I wouldn't use this to influence decisions.

Both issues would be fixed by setting up a private pip server and a mechanism to push arbitrarily to it. But since there's no such server currently, the solution in this PR addresses the issues.

AurelienGasser commented:
@ClementGautier I'm addressing your other suggestions, thanks!

AurelienGasser commented:
@samlesu Sounds like a great solution, let me give it a shot

samlesu commented on Makefile:
```diff
@@ -10,3 +18,15 @@ test-minimal: pyclean
+#   make docker DOCKER_TAG=master SUBSTRA_GIT_REF=master
+#   make docker DOCKER_TAG=1.0    SUBSTRA_GIT_REF=v1.0
+#   make docker DOCKER_TAG=pr_123 SUBSTRA_GIT_REF=refs/pull/123/head
+#   make docker DOCKER_TAG=commit SUBSTRA_GIT_REF@da39a3ee5e6b4b0d3255bfef95601890afd80709
```
```suggestion
#   make docker DOCKER_TAG=commit SUBSTRA_GIT_REF=da39a3ee5e6b4b0d3255bfef95601890afd80709
```
Typo

AurelienGasser commented:
@ClementGautier any thoughts/improvement ideas?
